crates/editor/src/items.rs

```diff
@@ -1481,6 +1482,43 @@ pub fn entry_label_color(selected: bool) -> Color {
     }
 }
 
+pub fn entry_diagnostic_and_git_aware_extra_icon_name(
+    diagnostic_severity: Option<&DiagnosticSeverity>,
+    git_status: Option<GitFileStatus>,
+    ignored: bool,
+) -> Option<IconName> {
+    if let Some(diagnostic_severity) = diagnostic_severity {
+        if *diagnostic_severity == DiagnosticSeverity::ERROR {
+            return Some(IconName::XCircle);
+        }
+        if *diagnostic_severity == DiagnosticSeverity::WARNING {
+            return Some(IconName::Warning);
+        }
+    }
+    if !ignored && git_status.is_some() {
+        Some(IconName::Dot)
+    } else {
+        None
+    }
```

Comment:
Nitpick, but I'd use a `match` statement:

```suggestion
    match diagnostic_severity {
        Some(&DiagnosticSeverity::ERROR) => Some(IconName::XCircle),
        Some(&DiagnosticSeverity::WARNING) => Some(IconName::Warning),
        _ => {
            if !ignored && git_status.is_some() {
                Some(IconName::Dot)
            } else {
                None
            }
        }
    }
```